### PR TITLE
refactor(types): minor changes

### PR DESCRIPTION
**Pull request by rafaelss95**
Some minor changes to the types. I've separate it in multiple commits to make the reviewer's life easier :)

1st. commit: I've grouped types and interfaces separately in hope of improving readability. As a contributor, it caused some confusion for me in this when I read at the initial times, so I'm proposing this change.
2nd. commit: I saw that `hideClass` and `showClass` are exactly the same, so I created a separated interface to reuse it in them.
3rd. commit: Make `SweetAlertResult` members readonly to represent that they're immutable.
4th. commit: Some leftovers from #1983 (support readonly user inputs).
~5th. commit: Make **non** updatable parameters readonly. Refer to https://github.com/sweetalert2/sweetalert2/pull/1996#issuecomment-640766402 and previous discussions.~
6th. commit: Change the signature of `isUpdatableParam` and `isValidParameter` to be user-defined type guards, so we can properly check any string as ngx-sweetalert2 currently does in `SwalComponent`.

Note that:

1. The fact that there are multiple commits doesn't imply that you have to merge them all. I'm also fine with squash.
2. if any of the changes are not desirable I can remove it without problems.

cc @limonte @toverux

### CODE REVIEW

**Comment by rafaelss95** [comment]
Are there any interest in having tests for this using a tool like https://github.com/Microsoft/dtslint?

File: sweetalert2.d.ts [hunk 1]
```diff
@@ -422,7 +401,26 @@ declare module 'sweetalert2' {
     | 'title'
     | 'titleText';
 
-  export type SweetAlertArrayOptions = [string?, string?, SweetAlertIcon?];
+  export interface SweetAlertCustomClass extends SweetAlertCommonClass {
```
Comment (rafaelss95) [inline sweetalert2.d.ts:404]: A question that came to my mind when I was working on this... does `update` support "patch" update?

For example...

```ts
swal.fire({
  title: 'blabla',
  customClass: {
    popup: 'my-popup'
  }
})

swal.update({
  customClass: {
    backdrop: 'my-backdrop'
  }
})
```

... the `customClass` will be `{ popup: 'my-popup', backdrop: 'my-backdrop' }` or just `{ backdrop: 'my-backdrop' }`?

Comment (limonte) [reply sweetalert2.d.ts:404]: No, "patch" updates aren't supported.

File: sweetalert2.d.ts [hunk 2]
```diff
@@ -336,53 +336,13 @@ declare module 'sweetalert2' {
...
-    closeButton?: string;
-    icon?: string;
-    image?: string;
-    content?: string;
-    input?: string;
-    actions?: string;
-    confirmButton?: string;
-    cancelButton?: string;
-    footer?: string;
+  interface SweetAlertHideShowClass extends SweetAlertCommonClass {
```
Comment (limonte) [inline sweetalert2.d.ts:344]: `SweetAlertCommonClass` is an over-kill here, please remove it as it makes code less readable.

Comment (rafaelss95) [reply sweetalert2.d.ts:344]: > `SweetAlertCommonClass` is an over-kill here, please remove it as it makes code less readable.

Hmm, ok... but did you notice that I used it in `SweetAlertHideShowClass` and `SweetAlertCustomClass`?

**Comment by limonte** [comment]
> Are there any interest in having tests for this file using a tool like https://github.com/Microsoft/dtslint?

I believe `eslint` + `typescript-eslint` are doing well enough here, no need in adding `dtslint`.

**Comment by rafaelss95** [comment]
> > Are there any interest in having tests for this file using a tool like https://github.com/Microsoft/dtslint?
> 
> I believe `eslint` + `typescript-eslint` are doing well enough here, no need in adding `dtslint`.

Excuse me, _dtslint_ is a tool to test declaration files (it does other things, but in this case I'm referring to unit tests). It isn't a linter.

File: sweetalert2.d.ts (hunk 2 above)
Comment (limonte) [reply sweetalert2.d.ts:344]: Yes, https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

**Comment by limonte** [comment]
> > > Are there any interest in having tests for this file using a tool like https://github.com/Microsoft/dtslint?
> > 
> > 
> > I believe `eslint` + `typescript-eslint` are doing well enough here, no need in adding `dtslint`.
> 
> Excuse me, _dtslint_ is a tool to test declaration files (it does other things, but in this case I'm referring to unit tests). It isn't a linter.

Oops, sorry, let me check more carefully what dtslint is.

File: sweetalert2.d.ts (hunk 2 above)
Comment (rafaelss95) [reply sweetalert2.d.ts]: Thanks for the link :) I sent a fixup for that now.

**Comment by rafaelss95** [comment]
> This change results into:
> 
> image
> 
> on the side of our Angular integration.

Hmm, that's the same case I answered [**here**](https://github.com/sweetalert2/sweetalert2/pull/1983#issuecomment-640290307). What should I do? Remove the last commit that prevent mutations or send a PR in the ngx-sweetalert2 with the changes I proposed in the link above?

File: sweetalert2.d.ts (hunk 1 above)
Comment (zenflow) [reply sweetalert2.d.ts]: @rafaelss95 The `update` method doesn't do "patch" updates by itself, but you can accomplish this using the `params` member. Example:

```js
const swal = Swal.fire({
  customClass: { popup: 'my-popup' }
})
swal.update({
  customClass: { ...swal.params, backdrop: 'my-backdrop' }
})
console.log(swal.params.customClass) // { popup: 'my-popup', backdrop: 'my-backdrop' }
```

**Comment by toverux** [comment]
> 5th. commit: Make non updatable parameters readonly. Refer to #1996 (comment) and previous discussions.

Why only non-updatable parameters?
If I fire a swal, then mutate the options object directly, that won't update the modal, so why should they be declared differently?

Otherwise, good work, lgtm.

**Comment by rafaelss95** [comment]
> > 5th. commit: Make non updatable parameters readonly. Refer to #1996 (comment) and previous discussions.
> 
> Why only non-updatable parameters?
> If I fire a swal, then mutate the options object directly, that won't update the modal, so why should they be declared differently?
> 
> Otherwise, good work, lgtm.

I decided to drop the 5th commit as the use-case you described in #1983 is more clear now (I thought you're talking about having a open swal).

<hr>

@limonte I think it's ready now.

File: sweetalert2.d.ts (hunk 1 above)
Comment (rafaelss95) [reply sweetalert2.d.ts]: @zenflow Yes, I'm aware of that way :)

**Pull request merged**